app/src/main/java/com/github/libretube/services/DownloadService.kt

```diff
@@ -1,182 +1,524 @@
+    private val downloadQueue = mutableMapOf<Int, Boolean>()
+    private val _downloadFlow = MutableSharedFlow<Pair<Int, DownloadStatus>>()
+    val downloadFlow: SharedFlow<Pair<Int, DownloadStatus>> = _downloadFlow
 
     override fun onCreate() {
         super.onCreate()
         IS_DOWNLOAD_RUNNING = true
+        notifyForeground()
+        sendBroadcast(Intent(ACTION_SERVICE_STARTED))
     }
 
     override fun onStartCommand(intent: Intent?, flags: Int, startId: Int): Int {
-        videoName = intent?.getStringExtra("videoName")!!
-        videoUrl = intent.getStringExtra("videoUrl")!!
-        audioUrl = intent.getStringExtra("audioUrl")!!
-
-        downloadType = when {
-            videoUrl != "" && audioUrl != "" -> DownloadType.AUDIO_VIDEO
-            audioUrl != "" -> DownloadType.AUDIO
-            videoUrl != "" -> DownloadType.VIDEO
-            else -> DownloadType.NONE
+        when (intent?.action) {
+            ACTION_DOWNLOAD_RESUME -> resume(intent.getIntExtra("id", -1))
+            ACTION_DOWNLOAD_PAUSE -> pause(intent.getIntExtra("id", -1))
         }
 
-        if (downloadType != DownloadType.NONE) {
-            downloadManager()
-        } else {
-            onDestroy()
+        val videoId = intent?.getStringExtra(IntentData.videoId) ?: return START_NOT_STICKY
+        val fileName = intent.getStringExtra(IntentData.fileName) ?: videoId
+        val videoFormat = intent.getStringExtra(IntentData.videoFormat)
+        val videoQuality = intent.getStringExtra(IntentData.videoQuality)
+        val audioFormat = intent.getStringExtra(IntentData.audioFormat)
+        val audioQuality = intent.getStringExtra(IntentData.audioQuality)
+        val subtitleCode = intent.getStringExtra(IntentData.subtitleCode)
+
+        scope.launch {
+            try {
+                val streams = RetrofitInstance.api.getStreams(videoId)
+
+                awaitQuery {
+                    Database.downloadDao().insertDownload(
+                        Download(
+                            videoId = videoId,
+                            title = streams.title ?: "",
+                            thumbnailPath = File(
+                                DownloadHelper.getDownloadDir(
+                                    this@DownloadService,
+                                    DownloadHelper.THUMBNAIL_DIR
+                                ),
+                                fileName
+                            ).absolutePath,
+                            description = streams.description ?: "",
+                            uploadDate = streams.uploadDate,
+                            uploader = streams.uploader ?: ""
+                        )
+                    )
+                }
+                streams.thumbnailUrl?.let { url ->
+                    ImageHelper.downloadImage(
+                        this@DownloadService,
+                        url,
+                        File(
```

Comment:
We should create a function that returns the download directory only based on one argument like DownloadHelper#THUMBNAIL_DIR  in order to not repeat that code for creating a file multiple times.

Reply:
Not sure wdym. current `getDownloadDir` takes two args that are required.

Reply:
Always when calling
```kotlin
File(DownloadHelper.getDownloadDir(), fileName)
```
here in the DownloadService, the `fileName` and the `context` parameter are always the name. Hence I suggest to create a function that only takes the `path` parameter of `DownloadHelper#getDownloadDir` and creates a file as mentioned above from it. Hope that's understandable, if not feel free to ask :)

Reply:
Alright, i will try that.